test/validation/api/ipsec/ipsec_test_out.c

```diff
@@ -407,22 +419,44 @@ static void test_out_in_common(ipsec_test_flags *flags,
 	    (auth == ODP_AUTH_ALG_NULL))
 		return;
 
+	if (flags->v4_tunnel) {
+		memset(&tunnel, 0, sizeof(odp_ipsec_tunnel_param_t));
+		tunnel.type = ODP_IPSEC_TUNNEL_IPV4;
+		tunnel.ipv4.src_addr = &src_v4;
+		tunnel.ipv4.dst_addr = &dst_v4;
+		tun_ptr = &tunnel;
+	}
+	if (flags->v6_tunnel) {
+		memset(&tunnel, 0, sizeof(odp_ipsec_tunnel_param_t));
+		tunnel.type = ODP_IPSEC_TUNNEL_IPV6;
+		tunnel.ipv6.src_addr = &src_v6;
+		tunnel.ipv6.dst_addr = &dst_v6;
+		tunnel.ipv6.hlimit = 64;
+		tun_ptr = &tunnel;
+	}
+
 	ipsec_sa_param_fill(&param,
-			    false, flags->ah, 123, NULL,
+			    false, flags->ah, 123, tun_ptr,
 			    cipher, cipher_key,
 			    auth, auth_key,
 			    cipher_key_extra, auth_key_extra);
 
+	if (flags->udp_encap)
+		param.opt.udp_encap = 1;
+
 	sa_out = odp_ipsec_sa_create(&param);
 
 	CU_ASSERT_NOT_EQUAL_FATAL(ODP_IPSEC_SA_INVALID, sa_out);
 
 	ipsec_sa_param_fill(&param,
-			    true, flags->ah, 123, NULL,
+			    true, flags->ah, 123, tun_ptr,
 			    cipher, cipher_key,
 			    auth, auth_key,
 			    cipher_key_extra, auth_key_extra);
 
+	if (flags->udp_encap)
+		param.opt.udp_encap = 1;
```

Comment:
Just setting this parameter and then checking that IPsec transformations still work does not test that UDP encapsulation is even done. An implementation could simply ignore the parameter and perform plain non-UDP encapsulated transformations and this validation test would not notice anything wrong.

So, I think at minimum some additional checks are needed to see that UDP encapsulation was actually done, even if its correctness is not fully tested.

Reply:
I'll add an assert for ipv4.proto & ipv6.next_hdr. Hope that would suffice.

Reply:
That's enough for me now. It can be improved later if needed but this change already improves test coverage.